From 7aead4218954ea14c7a03ac1e3deaf627996c3d1 Mon Sep 17 00:00:00 2001
From: Wei-Chiu Chuang <weichiu@cloudera.com>
Date: Mon, 2 May 2016 06:43:03 -0700
Subject: [PATCH 1482/2748] HDFS-10182. Hedged read might overwrite user's
 buf. Contributed by zhouyingchao.

(cherry picked from commit 173681a477309ca3561d3dc36eade9bb9af9fcbc)

Change-Id: If8922ad15b73361581c9871bcf71e1f43c174a58
---
 .../org/apache/hadoop/hdfs/DFSInputStream.java     |   16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java
index af66b7c..7760e3b 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java
@@ -1268,7 +1268,7 @@ private void hedgedFetchBlockByteRange(LocatedBlock block, long start,
         // chooseDataNode is a commitment. If no node, we go to
         // the NN to reget block locations. Only go here on first read.
         chosenNode = chooseDataNode(block, ignored);
-        bb = ByteBuffer.wrap(buf, offset, len);
+        bb = ByteBuffer.allocate(len);
         Callable<ByteBuffer> getFromDataNodeCallable = getFromOneDataNode(
             chosenNode, block, start, end, bb, corruptedBlockMap,
             hedgedReadId++);
@@ -1279,7 +1279,9 @@ private void hedgedFetchBlockByteRange(LocatedBlock block, long start,
           Future<ByteBuffer> future = hedgedService.poll(
               dfsClient.getHedgedReadTimeout(), TimeUnit.MILLISECONDS);
           if (future != null) {
-            future.get();
+            ByteBuffer result = future.get();
+            System.arraycopy(result.array(), result.position(), buf, offset,
+                len);
             return;
           }
           if (DFSClient.LOG.isDebugEnabled()) {
@@ -1325,13 +1327,9 @@ private void hedgedFetchBlockByteRange(LocatedBlock block, long start,
           ByteBuffer result = getFirstToComplete(hedgedService, futures);
           // cancel the rest.
           cancelAll(futures);
-          if (result.array() != buf) { // compare the array pointers
-            dfsClient.getHedgedReadMetrics().incHedgedReadWins();
-            System.arraycopy(result.array(), result.position(), buf, offset,
-                len);
-          } else {
-            dfsClient.getHedgedReadMetrics().incHedgedReadOps();
-          }
+          dfsClient.getHedgedReadMetrics().incHedgedReadWins();
+          System.arraycopy(result.array(), result.position(), buf, offset,
+              len);
           return;
         } catch (InterruptedException ie) {
           // Ignore and retry
-- 
1.7.9.5

